common/X11R5/clients/xbiff/xbiff.man
.\"ident	"@(#)r5xbiff:xbiff.man	1.1"
.TH XBIFF 1 "Release 5" "X Version 11"
.SH NAME
xbiff - mailbox flag for X
.SH SYNOPSIS
.B xbiff
[-\fItoolkitoption\fP ...] [-option ...]
.SH DESCRIPTION
The
.I xbiff
program displays a little image of a mailbox.  When there is no mail, the flag
on the mailbox is down.  When mail arrives, the flag goes up and the mailbox beeps.
By default,
pressing any mouse button in the image forces \fIxbiff\fP to remember the
current size of the mail file as being the ``empty'' size and to lower
the flag.
.PP
This program is nothing more than a wrapper around the Athena \fIMailbox\fP
widget.
.SH OPTIONS
.I Xbiff
accepts all of the standard X Toolkit command line options along with the
additional options listed below:
.TP 8
.B \-help
This option indicates that a brief summary of the allowed options should be
printed on the standard error.
.TP 8
.B \-update \fIseconds\fP
This option specifies the frequency in seconds at which \fIxbiff\fP
should update its display.  If the mailbox is obscured and then exposed,
it will be updated immediately.  The default is 30 seconds.
.TP 8
.B \-file \fIfilename\fP
This option specifies the name of the file which should be monitored.  By
default, it watches /usr/spool/mail/\fIusername\fP, where \fIusername\fP
is your login name.
.TP 8
.B \-volume \fIpercentage\fP
This option specifies how loud the bell should be rung when new mail comes in.
.TP 8
.B \-shape
This option indicates that the mailbox window should be shaped if masks for
the empty or full images are given.
.PP
The following standard X Toolkit command line arguments are commonly used with 
.I xbiff:
.TP 8
.B \-display \fIdisplay\fP
This option specifies the X server to contact.
.TP 8
.B \-geometry \fIgeometry\fP
This option specifies the preferred size and position of the mailbox window.
The mailbox is 48 pixels wide and 48 pixels high and will be centered in
the window.
.TP 8
.B \-bg \fIcolor\fP
This option specifies the color to use for the background of the window.  
.TP 8
.B \-bd \fIcolor\fP
This option specifies the color to use for the border of the window.
.TP 8
.B \-bw \fInumber\fP
This option specifies the width in pixels of the border surrounding the window.
.TP 8
.B \-fg \fIcolor\fP
This option specifies the color to use for the foreground of the window.
.TP 8
.B \-rv
This option indicates that reverse video should be simulated by swapping
the foreground and background colors.
.TP 8
.B \-xrm \fIresourcestring\fP
This option specifies a resource string to be used.  This is especially
useful for setting resources that do not have separate command line options.
.SH X DEFAULTS
The application class name is XBiff.
This program uses the 
.I Mailbox
widget in the Athena widget set.  It understands all of the core resource names and
classes as well as:
.PP
.TP 8
.B checkCommand (\fPclass\fB CheckCommand)
Specifies a shell command to be executed to check for new mail rather than
examining the size of \fBfile\fP.  The specified string value is used as the
argument to a \fIsystem\fP(3) call and may therefore contain i/o redirection.
An exit status of 0 indicates that new mail is waiting, 1 indicates that there
has been no change in size, and 2 indicates that the mail has been cleared.
By default, no shell command is provided.
.TP 8
.B file (\fPclass\fB File)
Specifies the name of the file to monitor.  The default is to watch
/usr/spool/mail/\fIusername\fP, where \fIusername\fP is your login name.
.TP 8
.B onceOnly (\fPclass\fB Boolean)
Specifies that the bell is only rung the first time new mail is found
and is not rung again until at least one interval has passed with
no mail waiting.  The window will continue to indicate the presence
of new mail until it has been retrieved.  The default is false.
.TP 8
.B width (\fPclass\fB Width)
Specifies the width of the mailbox.
.TP 8
.B height (\fPclass\fB Height)
Specifies the height of the mailbox.
.TP 8
.B update (\fPclass\fB Interval)
Specifies the frequency in seconds at which the mail should be checked.
The default is 30.
.TP 8
.B volume (\fPclass\fB Volume)
Specifies how loud the bell should be rung.  The default is 33 percent.
.TP 8
.B foreground (\fPclass\fB Foreground)
Specifies the color for the foreground.
.TP 8
.B reverseVideo (\fPclass\fB ReverseVideo)
Specifies that the foreground and background should be reversed.
.TP 8
.B flip (\fPclass\fB Flip)
Specifies whether or not the image that is shown when mail has arrived 
should be inverted.  The default is ``true.''
.TP 8
.B fullPixmap (\fPclass\fB Pixmap)
Specifies a bitmap to be shown when new mail has arrived.
The default is flagup.
.TP 8
.B emptyPixmap (\fPclass\fB Pixmap)
Specifies a bitmap to be shown when no new mail is present.
The default is flagdown.
.TP 8
.B shapeWindow (\fPclass\fB ShapeWindow)
Specifies whether or not the mailbox window should be shaped to the
given fullPixmapMask and emptyPixmapMask.  The default is false.
.TP 8
.B fullPixmapMask (\fPclass\fB PixmapMask)
Specifies a mask for the bitmap to be shown when new mail has arrived.
The default is none.
.TP 8
.B emptyPixmapMask (\fPclass\fB PixmapMask)
Specifies a mask for the bitmap to be shown when no new mail is present.
The default is none.
.SH ACTIONS
The \fIMailbox\fP widget provides the following actions for use in event
translations:
.TP 8
.B check()
This action causes the widget to check for new mail and display the flag
appropriately.
.TP 8
.B unset()
This action causes the widget to lower the flag until new mail comes in.
.TP 8
.B set()
This action causes the widget to raise the flag until the user resets it.
.PP
The default translation is 
.sp
.nf
        <ButtonPress>:  unset()
.fi
.sp
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH "SEE ALSO"
X(1),
xrdb(1),
stat(2)
.SH BUGS
The mailbox bitmaps are ugly.
.SH COPYRIGHT
Copyright 1988, Massachusetts Institute of Technology.
.br
See \fIX(1)\fP for a full statement of rights and permissions.
.SH AUTHOR
Jim Fulton, MIT X Consortium
.br
Additional hacks by Ralph Swick, DEC/MIT Project Athena


common/X11R5/clients/xbiff/xbiff.c
/*	Copyright (c) 1990, 1991, 1992 UNIX System Laboratories, Inc.	*/
/*	Copyright (c) 1988, 1989, 1990 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF     	*/
/*	UNIX System Laboratories, Inc.                     	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)r5xbiff:xbiff.c	1.1"
/*
 * $XConsortium: xbiff.c,v 1.17 91/01/10 14:25:32 gildea Exp $
 *
 * Copyright 1988 Massachusetts Institute of Technology
 *
 *
 * Author:  Jim Fulton, MIT X Consortium
 */

#include <stdio.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xaw/Mailbox.h>
#include <X11/Xaw/Cardinals.h>

extern void exit();
static void quit();

char *ProgramName;

static XrmOptionDescRec options[] = {
{ "-update", "*mailbox.update", XrmoptionSepArg, (caddr_t) NULL },
{ "-file",   "*mailbox.file", XrmoptionSepArg, (caddr_t) NULL },
{ "-volume", "*mailbox.volume", XrmoptionSepArg, (caddr_t) NULL },
{ "-shape",  "*mailbox.shapeWindow", XrmoptionNoArg, (caddr_t) "on" },
};

static XtActionsRec xbiff_actions[] = {
    { "quit", quit },
};
static Atom wm_delete_window;

static void Usage ()
{
    static char *help_message[] = {
"where options include:",
"    -display host:dpy              X server to contact",
"    -geometry geom                 size of mailbox",
"    -file file                     file to watch",
"    -update seconds                how often to check for mail",
"    -volume percentage             how loud to ring the bell",
"    -bg color                      background color",
"    -fg color                      foreground color",
"    -rv                            reverse video",
"    -shape                         shape the window",
NULL};
    char **cpp;

    fprintf (stderr, "usage:  %s [-options ...]\n", ProgramName);
    for (cpp = help_message; *cpp; cpp++) {
	fprintf (stderr, "%s\n", *cpp);
    }
    fprintf (stderr, "\n");
    exit (1);
}


void main (argc, argv)
    int argc;
    char **argv;
{
    XtAppContext xtcontext;
    Widget toplevel, w;

    ProgramName = argv[0];

    toplevel = XtAppInitialize(&xtcontext, "XBiff", options, XtNumber (options),
			       &argc, argv, NULL, NULL, 0);
    if (argc != 1) Usage ();

    /*
     * This is a hack so that f.delete will do something useful in this
     * single-window application.
     */
    XtAppAddActions (xtcontext, xbiff_actions, XtNumber(xbiff_actions));
    XtOverrideTranslations(toplevel,
		   XtParseTranslationTable ("<Message>WM_PROTOCOLS: quit()"));

    w = XtCreateManagedWidget ("mailbox", mailboxWidgetClass, toplevel,
			       NULL, 0);
    XtRealizeWidget (toplevel);
    wm_delete_window = XInternAtom (XtDisplay(toplevel), "WM_DELETE_WINDOW",
                                    False);
    (void) XSetWMProtocols (XtDisplay(toplevel), XtWindow(toplevel),
                            &wm_delete_window, 1);
    XtAppMainLoop (xtcontext);
}

static void quit (w, event, params, num_params)
    Widget w;
    XEvent *event;
    String *params;
    Cardinal *num_params;
{
    if (event->type == ClientMessage &&
        event->xclient.data.l[0] != wm_delete_window) {
        XBell (XtDisplay(w), 0);
        return;
    }
    XCloseDisplay (XtDisplay(w));
    exit (0);
}
